### ISSUE TITLE
serde::de::Visitor's Value bound of Deserialize, now for template structs

### ISSUE THREAD

**Issue by 0x7CFE** [serde-rs/serde#815]
**tl;dr:** much like #330 but for structs with template parameters (I think).

I tried to serialize the following struct:
```rust
#[derive(Clone, Hash, Eq, PartialEq, Serialize, Deserialize)]
pub struct SparseBitVec {
    leading_zeros: usize,
    trailing_zeros: usize,
    bits_set: usize,

    #[serde(serialize_with = "serialize_bits", deserialize_with = "deserialize_bits")]
    bits: BitVec,
}
```

Since `BitVec` (from the [bit_vec crate](https://crates.io/crates/bit-vec)) does not support serde yet, I must implement it on my own. AFAIK there are two ways to do that: either using the newtype pattern, or using `[de]serialize_with` attributes on a field.

Newtype works OK, but it's not very convenient for me so I tried to use the field attribute. I implemented the functions:

```rust

fn serialize_bits<S>(bits: &BitVec, serializer: &mut S) -> Result<(), S::Error>
    where S: Serializer
{
    let bytes = bits.to_bytes();
    serializer.serialize_bytes(&bytes)
}

fn deserialize_bits<D>(deserializer: &mut D) -> Result<BitVec, D::Error>
    where D: Deserializer
{
    struct BitVecVisitor;

    impl Visitor for BitVecVisitor {
        type Value = BitVec;

        fn visit_seq<V>(&mut self, mut visitor: V) -> Result<Self::Value, V::Error>
            where V: SeqVisitor
        {
            let len = visitor.size_hint().0;
            let mut bytes = Vec::with_capacity(len);
            while let Some(value) = visitor.visit()? {
                bytes.push(value);
            }
        
            Ok(BitVec::from_bytes(&bytes))
        }

        #[inline]
        fn visit_bytes<E>(&mut self, v: &[u8]) -> Result<Self::Value, E>
            where E: Error
        {
            Ok(BitVec::from_bytes(v))
        }
    }

    deserializer.deserialize_bytes(BitVecVisitor)
}
```

Unfortunately now compiler complains:
```
error[E0277]: the trait bound `memory::BitVec: serde::Deserialize` is not satisfied                                                                                                                                      
   --> src/memory.rs:105:10                                                                                                                                                                                              
    |                                                                                                                                                                                                                    
105 |     impl Visitor for BitVecVisitor {                                                                                                                                                                               
    |          ^^^^^^^ the trait `serde::Deserialize` is not implemented for `memory::BitVec`                                                                                                                            
    |                                                                                                                                                                                                                    
    = note: required by `serde::de::Visitor`                                                                                                                                                                             
                                                                                                                                                                                                                         
error: aborting due to previous error                                                                                                                                                                                    
```
I thought that attributes are now fixed, but looks like they are not.

FYI: `BitVec` is defined as:
```rust
pub struct BitVec<B=u32> {
    /// Internal representation of the bit vector
    storage: Vec<B>,
    /// The number of valid bits in the internal representation
    nbits: usize
}
```

Related version info:

```
serde = "0.8"
serde_derive = "0.8"
bincode = "0.6"
```

```
rustc 1.17.0-nightly (b1e31766d 2017-03-03)
binary: rustc
commit-hash: b1e31766da75b188062c59f38cd6e8544b902afd
commit-date: 2017-03-03
host: x86_64-unknown-linux-gnu
release: 1.17.0-nightly
LLVM version: 3.9
```

**Comment by dtolnay** [comment]
This was fixed in Serde 0.9.0. By the way, also check out the `with` attribute which would make your struct definition a bit more readable:

```rust
#[derive(Clone, Hash, Eq, PartialEq, Serialize, Deserialize)]
pub struct SparseBitVec {
    leading_zeros: usize,
    trailing_zeros: usize,
    bits_set: usize,

    #[serde(with = "bits")]
    bits: BitVec,
}

mod bits {
    fn serialize<S>(v: &BitVec, serializer: S) -> Result<S::Ok, S::Error>
        where S: Serializer
    {
        /* ... */
    }

    fn deserialize<D>(deserializer: D) -> Result<BitVec, D::Error>
        where D: Deserializer
    {
        /* ... */
    }
}
```

**Comment by 0x7CFE** [comment]
Oh, I'll try it. Thank you!
